### PR TITLE
Rework prepared queries

### PR DESCRIPTION
**Pull request by vietj**
fixes 

- https://github.com/eclipse-vertx/vertx-sql-client/issues/675
- https://github.com/eclipse-vertx/vertx-sql-client/issues/682
- https://github.com/eclipse-vertx/vertx-sql-client/issues/685

### CODE REVIEW

**Comment by BillyYccc** [comment]
Does this imply the statement auto-closing after one-shot preparedQuery is handled by the codec layer?

**Comment by vietj** [comment]
@BillyYccc yes it is implied

**Comment by vietj** [comment]
@BillyYccc @aguibert I would like to know if the above is fine and then we can start finishing this PR together and after back-port it to 3.9

**Comment by vietj** [comment]
It might be possible actually to avoid changes in the codec by moving the two step execution logic in the socket instead of moving to the codec which would avoid changing code in the current implementations of the codecs, I'll keep you informed of the progress.

**Comment by vietj** [comment]
I updated this PR with an intermediate change that does not require modification of current codecs and is actually more satisfying. I'll update the proposal issue to be in sync with this change.

**Comment by vietj** [comment]
so this should not involve contributions from @aguibert and @BillyYccc

**Comment by vietj** [comment]
Updated with prepared query caching simplification and corrections.

**Comment by vietj** [comment]
I'm moving this PR to review

**Comment by vietj** [comment]
To ease the review I will push the first commit that refactor the extended query commands in a single command to master as it is not related to the PR and the goal was to simplify and help this change

**Comment by vietj** [comment]
this is ready for review

**Comment by vietj** [comment]
Pushed a fix for the incorrect pipelining count that was found by a case provided by @BillyYccc

File: vertx-sql-client/src/main/java/io/vertx/sqlclient/impl/SocketConnectionBase.java [hunk 1]
```diff
@@ -70,7 +72,8 @@ public SocketConnectionBase(NetSocketInternal socket,
     this.socket = socket;
     this.context = context;
     this.pipeliningLimit = pipeliningLimit;
-    this.psCache = cachePreparedStatements ? new PreparedStatementCache(this, preparedStatementCacheSize) : null;
+    this.paused = false;
+    this.psCache = cachePreparedStatements ? new PreparedStatementCache(preparedStatementCacheSize) : null;
     this.preparedStatementCacheSqlLimit = preparedStatementCacheSqlLimit;
```
Comment (BillyYccc) [inline vertx-sql-client/src/main/java/io/vertx/sqlclient/impl/SocketConnectionBase.java:77]: we should use this option in the cache to make it work like before

Comment (BillyYccc) [reply vertx-sql-client/src/main/java/io/vertx/sqlclient/impl/SocketConnectionBase.java:77]: I mean `preparedStatementCacheSqlLimit` here

Comment (vietj) [reply vertx-sql-client/src/main/java/io/vertx/sqlclient/impl/SocketConnectionBase.java:77]: why is this feature useful ? can we just instead remove it ?

**Comment by vietj** [comment]
I back ported the commits @BillyYccc

File: vertx-sql-client/src/main/java/io/vertx/sqlclient/impl/SocketConnectionBase.java (hunk 1 above)
Comment (vietj) [reply vertx-sql-client/src/main/java/io/vertx/sqlclient/impl/SocketConnectionBase.java]: I added a generic filter for this, see the latest commit

**Comment by vietj** [comment]
@BillyYccc added support for SQL cache filtering, I think now we are good and we can do a final pass on the review

**Comment by vietj** [comment]
it was a collaboration as usual @BillyYccc :-)

**Pull request merged**